Sources/NIOSSL/SSLContext.swift

```diff
@@ -520,6 +527,83 @@ extension NIOSSLContext {
     }
 }
 
+// For utilizing SNI callbacks from a SSLContext
+extension NIOSSLContext {
+    
+    private static func setSniCallback(context: OpaquePointer) throws {
+        
+        CNIOBoringSSL_SSL_CTX_set_tlsext_servername_callback(context) { (ssl, adPointer, argPointer) in
+            guard let ssl = ssl else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Make sure we get the server name extracted.
+            guard let servername = CNIOBoringSSL_SSL_get_servername(ssl, CNIOBoringSSL.TLSEXT_NAMETYPE_host_name) else
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            let serverNameString = String(cString: servername)
+            
+            // Can do nothing about empty indicators.
+            if (serverNameString.isEmpty)
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // We want to take the SSL pointer and extract the parent Swift object. These force-unwraps are for
+            // safety: a correct NIO program can never fail to set these pointers, and if it does failing loudly is
+            // more useful than failing quietly.
+            let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+            let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+            let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
```

Comment:
This pattern is common enough that I think we should probably extract it into a helper function now.

Reply:
Agreed. I copy and pasted existing code.